posts/2021/01/enhancements-to-the-guvectorize-decorator.md

```diff
@@ -0,0 +1,150 @@
+<!--
+.. title: Enhancements to the guvectorize decorator
+.. slug: enhancements-to-the-guvectorize-decorator
+.. date: 2021-01-20 08:00:00 UTC-00:00
+.. author: Guilherme Leobas
+.. tags: Labs, Numba
+.. category:
+.. link:
+.. description:
+.. type: text
+-->
+
+Starting on Numba 0.53, Numba will ship with an enhanced version of the `@guvectorize` decorator. Similar to the [@vectorize](https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-vectorize-decorator) decorator, [@guvectorize](https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-guvectorize-decorator) now has two modes of operation: 
+
+- Eager, or decoration-time compilation and
+- Lazy, or call-time compilation
+
+Before, only the eager approach was supported. In this mode, users are required to provide a list of concrete supported types beforehand as its first argument. Now, this list can be omitted if desired and as one calls it, Numba dynamically generates new kernels for previously unsupported types.
+
+<!-- TEASER_END -->
+
+## NumPy Universal Functions
+
+NumPy introduces a type of function called Universal functions or ufuncs. Ufuncs are functions that operate on `ndarrays` element-by-element. [Examples](https://numpy.org/doc/stable/reference/ufuncs.html#available-ufuncs) of universal functions are `np.log` and `np.log2`, which compute the natural and base-2 logarithms, respectively. Alongside ufuncs, NumPy also introduces the notion of generalized ufuncs or gufuncs. While the former is limited to element-by-element operations, the latter supports sub-array by sub-array operations at a time.
+
+Creating new NumPy ufuncs is not an easy process and may require one to [write some C code](https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html). Numba extends the NumPy mechanism for registering and using (generalized) universal functions with two decorators: `@vectorize` and `@guvectorize`. Those decorators allow one to easily create universal functions from Python, leaving the gruntwork to Numba.
+
+For instance, consider the function `guvec`, which adds a scalar to every element in an array:
+
+```python
+# examples taken from https://numba.pydata.org/numba-doc/latest/user/vectorize.html
+from numba import guvectorize, int64
+import numpy as np
+
+@guvectorize([(int64[:], int64, int64[:])], '(n),()->(n)')
+def guvec(x, y, res):
+    for i in range(x.shape[0]):
+        res[i] = x[i] + y
```

Comment:
I guess this is from the NumPy docs, but I don't understand this example. Isn't this something that could be done with @vectorize as well?

Reply:
Actually, this example is from the [numba docs](https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-guvectorize-decorator). And yes, I guess one could do the same with `@vectorize`.

Reply:
Maybe worth changing then?

Reply:
Yes, I meant numba, not numpy.

Reply:
I've updated the example

Reply:
@asmeurer can you double check the new example?